clinica/iotools/converters/oasis3_to_bids/oasis3_utils.py

```diff
@@ -77,28 +77,53 @@ def read_imaging_data(imaging_data_directory: PathLike) -> DataFrame:
     source_dir_series = source_path_series.apply(lambda x: Path(str(x)).parent).rename(
         "source_dir"
     )
-
     file_spec_series = source_path_series.apply(lambda x: Path(str(x)).parts[0]).rename(
         "path"
     )
-
+    source_file_series = source_path_series.apply(
+        lambda x: identify_modality(x)
+    ).rename("modality")
+    source_run_series = source_path_series.apply(
+        lambda x: identify_runs(str(x))
+    ).rename("run_number")
     df_source = pd.concat(
         [
             source_path_series,
             file_spec_series,
             source_dir_series,
             file_spec_series.str.split("_", expand=True),
+            source_file_series,
+            source_run_series,
         ],
         axis=1,
     )
-    df_source = df_source.rename(
-        {0: "Subject", 1: "modality", 2: "Date"}, axis="columns"
-    ).drop_duplicates()
-
+    df_source = (
+        df_source.rename({0: "Subject", 1: "modality_2", 2: "Date"}, axis="columns")
+        .drop_duplicates()
+        .sort_values(by=["source_path"])
+    )
     df_source = df_source.assign(participant_id=lambda df: "sub-" + df.Subject)
+    df_source["modality"] = df_source[["modality", "modality_2"]].apply(
+        "_".join, axis=1
+    )
     return df_source
 
 
+def identify_modality(x: str) -> str:
+    from pathlib import Path
+
+    return (Path(str(x)).name).split(".")[0].split("_")[-1]
```

Comment:
This can fail for a lot of reasons if the input data isn't what you expect for some reason. 
I would have it in a try except and raise a meaningful error. WDYT ?